File: downloadablebookandexercisecode/1256/code_files/Chapter_07_Strings/USACO_2021_feb_b1_year.cpp
#include <iostream>

#include <string>

using namespace std;

int main(){

    int N;
    cin >> N;

    string years[]={"Ox", "Tiger", "Rabbit", "Dragon", "Snake",
                    "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig", "Rat"};
    string cow_name[N+1];
    int cow_age[N+1];
    int cow_year[N+1];

    cow_name[0]="Bessie";
    cow_age[0] = 0;
    cow_year[0] = 0;

    string cow_to, cow_from, tmp;
    string b4next, year_sign;
    for (int i=1; i<=N; ++i){
        cin >> cow_to >> tmp >> tmp >> b4next >> year_sign >> tmp >> tmp >> cow_from;
        cow_name[i] = cow_to;
        for (int j=0; j<12; ++j) {
            if (year_sign == years[j]) {
                cow_year[i] = j;
                break;
            }
        }

        int cow_from_ind;
        for (int j=0; j<i; ++j) {
            if (cow_from == cow_name[j]) {
                cow_from_ind = j;
                break;
            }
        }
        if (b4next == "next"){
            cow_age[i] = cow_age[cow_from_ind] + (cow_year[i] - cow_year[cow_from_ind] + 12) % 12;
            if (cow_year[i] == cow_year[cow_from_ind])
                cow_age[i] += 12;
        } else {
            cow_age[i] = cow_age[cow_from_ind] - (cow_year[cow_from_ind] - cow_year[i] + 12) % 12;
            if (cow_year[i] == cow_year[cow_from_ind])
                cow_age[i] -= 12;
        }

    }
    for (int j=0; j<N+1; ++j) {
        if ("Elsie" == cow_name[j]) {
            cout << abs(cow_age[j]) << endl ;
            break;
        }
    }

    return 0;
}

File: downloadablebookandexercisecode/1256/code_files/Chapter_07_Strings/USACO_2021_feb_b1_year.py
# http://www.usaco.org/index.php?page=viewproblem2&cpid=1107

N = int(input())

year = ['Ox', 'Tiger', 'Rabbit', 'Dragon', 'Snake', 'Horse', 'Goat', 'Monkey', 'Rooster', 'Dog', 'Pig', 'Rat'];
index = range(12)
year_dic = dict(zip(year, index))
cow_age = {'Bessie':0}
cow_year = {'Bessie':year_dic['Ox']}

for i in range(N):
    line = input()
    [cow_to, _, _, b4next, symbol, _, _, cow_from] = line.split()
    # Mildred born in next Dragon year from Bessie
    # print(f'cow_to={cow_to}, cow_from={cow_from}, b4next={b4next}')
    year_from = cow_year[cow_from]
    year_to = year_dic[symbol]
    if b4next=='previous':
        age = cow_age[cow_from] - (year_from - year_to + 12)%12
        if year_to == year_from:
            age -= 12
    if b4next == 'next':
        age = cow_age[cow_from] + (year_to - year_from + 12)%12
        if year_to == year_from:
            age += 12
    cow_year[cow_to] = year_to
    cow_age[cow_to] = age

print(f'{abs(cow_age["Elsie"])}')